// @core/build/config/babel-loader.js
// const resolve = require('@core/lib/project/resolve');

const plugins = [
  require('babel-plugin-syntax-dynamic-import'),
  require('babel-plugin-transform-object-rest-spread'),
  // require('babel-plugin-transform-runtime'), // @TODO should be enabled for performance but was causing issues
  require('babel-plugin-transform-remove-strict-mode'), // @TODO this is pretty bad. should be enabled and issues addressed
];

if (process.env.NODE_ENV === 'test') {
  plugins.push([require('babel-plugin-istanbul').default, {
    exclude: ['test*(s)/**'],
  }]);
}

module.exports = {
  // transpile all files that are not third party. Includes project source and child dependencies
  test: /\.js$/,
  include: [
    /node_modules\/(@core|@feature|@component|@page)/,
    /src\/.*js$/,
    /tests\/.*js$/,
    /app(-[A-Za-z0-9]+)?\.js$/,
  ],
  // exclude: /node_modules/,

  exclude: [
    /(node_modules\/(?!@core|@feature|@component|@page).)/,
  ],
  // test: /^((?!node_modules\/(?!@core|@feature|@component|@page).).)*\.js$/gm,
  loader: 'babel-loader',
  options: {
    presets: [require('babel-preset-env')],
    plugins,
  },
};
