// @component/product-thumbnail/app.js
import ObjectUtil from '@component/common/src/util/ObjectUtil';
import Marionette from 'backbone.marionette';
import Backbone from 'backbone';
import Logger from '@component/common/src/util/Logger';
import '@core/vendor/radio.shim';

const PageApp = new Marionette.Application();
PageApp.brand = 'bcom';
PageApp.on('start', () => {
  if (Backbone.history) {
    if (!Backbone.History.started) {
      Backbone.history.start({
        pushState: true,
        silent: false,
      });
    }
  }
});
try {
  PageApp.meta = {
    meta: {
      pageType: {
        search: true,
        subType: {
          dlp: true,
        },
      },
      properties: {
        local_killswitch_tree_path: false,
        facetPulseRateDelay: '400',
        seoDLPCopyblockEnabled: false,
        assetsHostUrl: 'https:assets.macysassets.com',
        gmClientId: 'gme-macysinc',
        gmeTableId: '06739517320133004748-11853667273131550346',
        HOST: 'http:localhost:2001',
        killswitch_json_modified_date: false,
        prosPanelEnabled: true,
        colorwayPricingEnabled: true,
        hideBOPSFacetEnabled: false,
        optimizedThumbnailEnabled: true,
        gmeAPIKey: 'AIzaSyCzwiHW1tSp_4FXaFuORRffbxBzQUN1qs4',
        brand: 'MCOM',
        hideRegPriceLabelEnabled: false,
      },
      context: {
        application: 'SITE',
        combination: '1121US',
        deviceType: 'DESKTOP',
        domestic: true,
        navigationType: 'SEARCH',
        regionCode: 'US',
        searchTerm: 'blue dresses',
        shoppingMode: 'SITE',
        currencyCode: 'USD',
        searchPass: 'exactMultiMatch',
      },
      analytics: {
        coremetrics: {
          clientID: '80067660',
          pageID: 'search_results',
          categoryID: 'DLP',
          searchResults: '801',
          attributes: [
            {
              name: 'colsPPP',
              value: '3|60',
              seq: '33',
            },
            {
              name: 'rdppInfo',
              value: '65',
              seq: '19',
            },
            {
              name: 'searchPassACorLS',
              value: 'Blue Dress | Search Pass:exactMultiMatch',
              seq: '22',
            },
          ],
        },
      },
      feature: {},
      component: {},
    },
  };
} catch (e) {
  Logger.log('Error:', e, 'parsing feature/recently-viewed data-bootstrap script block html as JSON.');
}

PageApp.getMeta = (key) => {
  let metaKey = key;
  if (key === undefined) {
    metaKey = 'meta';
  }
  return ObjectUtil.getChildProperty(PageApp.meta.meta, metaKey) || '';
};

export default PageApp;
