// @core/lib/webpack/plugins/progress.js
const ProgressBar = require('progress');
const chalk = require('chalk');

// this file requires webpack, but @core/lib does not include it directly as a dependency.
// this file should only be required in projects that have webpack (i.e. when using macys-cli or dev-server)
const webpack = require('webpack');

function ProgressPlugin() {
  const barLeft = chalk.bold('[');
  const barRight = chalk.bold(']');
  const preamble = chalk.cyan.bold('  build ') + barLeft;
  const barFormat = `${preamble}:bar${barRight}${chalk.green.bold(' :percent')}`;
  const stream = process.stdout;

  if (!stream.isTTY) {
    return () => {};
  }

  const barOptions = {
    complete: '=',
    incomplete: ' ',
    width: 20,
    total: 100,
    clear: true,
    stream,
  };

  const bar = new ProgressBar(barFormat, barOptions);

  let running = false;
  let startTime = 0;
  let lastPercent = 0;

  return new webpack.ProgressPlugin(((percent, msg) => {
    if (!running && lastPercent !== 0) {
      stream.write('\n');
    }

    const newPercent = Math.ceil(percent * barOptions.width);

    if (lastPercent !== newPercent) {
      bar.update(percent, { msg });
      lastPercent = newPercent;
    }

    if (!running) {
      running = true;
      startTime = new Date();
      lastPercent = 0;
    } else if (percent === 1) {
      const now = new Date();
      const buildTime = (now - startTime) / 1000;

      bar.terminate();

      // fast build times don't even render the bar
      if (buildTime > 0.1) {
        stream.write(chalk.green.bold(`Build completed in ${buildTime}s\n\n`));
      }

      running = false;
    }
  }));
}

module.exports = ProgressPlugin;
