// @component/common/src/components/backToTopButton/js/BackToTopButton.js
import $ from 'jquery';
import template from '../templates/backToTopButton.hbs';
import '../scss/back-to-top-button.scss';

function scrollToTop() {
  $('html,body').animate(
    {
      scrollTop: 0,
    },
    {
      duration: 400,
    },
  );
}

const WINDOW_SCROLL_EVENT = 'scroll.backToTopButton';
const BUTTON_CLICK_EVENT = 'click.backToTopButton';
const OFFSET_TO_APPEAR = 100;

export default class BackToTopButton {
  constructor(options = {}) {
    this.offsetToAppear = options.offsetToAppear || OFFSET_TO_APPEAR;

    $('body').append(template());

    this.$button = $('.back-to-top-button');
    this.updateButtonVisibility();

    $(window).on(WINDOW_SCROLL_EVENT, () => this.updateButtonVisibility());

    this.$button.on(BUTTON_CLICK_EVENT, (event) => {
      event.preventDefault();

      scrollToTop();
    });
  }

  updateButtonVisibility() {
    this.$button.toggleClass('show', $(window).scrollTop() > this.offsetToAppear);
  }

  destroy() {
    $(window).off(WINDOW_SCROLL_EVENT);

    this.$button.off(BUTTON_CLICK_EVENT);
    this.$button.remove();
  }
}
