// @component/common/src/util/BTDataDictionaryUtils.js
// #MODULE - BT Data Dictionary Utils
// > Author: Joseph Acosta
// >
// > Create Date: March 18, 2015
// >
// ##DESCRIPTION: A module that provides some base methods for manipulating the bright tag data dictionary
{
  const _ = require('underscore');
  function get(brand) {
    if (_.isUndefined(window[brand])) {
      window[brand] = {};
    }
    if (_.isUndefined(window[brand].brightTag)) {
      window[brand].brightTag = {};
    }
    return window[brand];
  }

  function clear(brand) {
    window[brand] = {
      brightTag: {},
    };
    const dd = window[brand];

    return dd;
  }

  function set(brand, key, value) {
    const dd = get(brand);

    dd.brightTag[key] = value;
    return dd;
  }

  function setNavigation(brand, pageName, pageId, keyword, registry) {
    const dd = get(brand);
    set(brand, 'navigation', {
      pageName,
      pageID: pageId,
      keywordSearch: keyword || '',
      registryMode: registry || 'NO',
    });
    return dd;
  }

  module.exports = {
    get,
    clear,
    set,
    setNavigation,
  };
}
