// @component/common/src/util/LazyImages.js
// #MODULE - Lazy Image Loading Utils
// > Author: Andrew Brassington
// >
// > Create Date: Nov 21, 2016
// >
// ##DESCRIPTION: A module that provides image lazy loading methods
// ##
// ##  Requires img elements to have .lazy class and data-src attribute with img src to be loaded (..ie):
// ##  <img class="lazy" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
// ##  data-src="//www.macys.com/web20/assets/img/nav/header-bottom-gradient.png" alt="Image Alt Text" border="0">
// ##
// ##  Recommended to use inline gif data img to maintain SEO on img alt text, prevents network call to load clear img
// ##
// ##  Example use in application; assign event handlers when initializing (also removes handler on event.target):
// ##     this.$el.find('li').one('mouseover.lazy click.lazy', LazyImages.loadContained);
// ##        OR invoke in Marionette events hash (does not clean up handler):
// ##     events: { 'click li': LazyImages.loadContained, 'mouseover li': LazyImages.loadContained},
// ##        OR invoke deferAllToLoadEvent before Load event fires:
// ##     LazyImages.deferAllToLoadEvent();
// ##
{
  const $ = require('jquery');
  function _imageLoader(element, complete, index) {
    const el = element;
    el.style.opacity = 0;
    el.style.transition = 'opacity 0.5s';
    const img = new Image();
    const src = el.getAttribute('data-src');
    img.onload = function () {
      el.src = src;
      el.style.opacity = 1;
      complete(el, index);
    };
    img.src = src;
    return el;
  }

  function _loadImages(target, callback) {
    let cb = callback;
    cb = typeof cb === 'function' ? cb : function () {};
    const untouchedImgs = Array.prototype.slice.call(target.querySelectorAll('img.lazy'));
    const loadedImgs = [];
    let loadedCount = 0;
    const completed = function (img, index) {
      loadedImgs[index] = img;
      loadedCount += 1;
      if (loadedCount === untouchedImgs.length) {
        cb(loadedImgs);
      }
    };
    return untouchedImgs.map((image, index) => _imageLoader(image, completed, index));
  }

  // ###Method - loadAll(event) method to load all images in document with .lazy class
  //
  // > parameters
  // >
  // + *event* - a native dom event
  // + *function* - optional callback( images ) invoked on image load completion with loaded images array
  //
  // > returns
  // >
  // + Array of image DOM elements being loaded (scheduled to load)
  function loadAll(event, cb) {
    const images = _loadImages(document, cb);
    return images;
  }

  // ###Method - loadContained(event) method to load all images with .lazy class contained within event.target
  //  Requires either event with event.target or element containing lazy images
  //  Lazy loads jquery to remove event listener on event.target
  //
  // > parameters
  // >
  // + *event* - a native dom event
  // + *element* - optional dom element containing lazy images to be loaded
  // + *function* - optional callback( images ) invoked on image load completion with loaded images array
  //
  // > returns
  // >
  // + Array of image DOM elements being loaded (scheduled to load)
  function loadContained(event, el, cb) {
    const target = el !== undefined ? el : event.target;
    if (target !== undefined) {
      const images = _loadImages(target, cb);
      $(target).off('.lazy'); // Removes listeners with jquery namespace: .lazy
      return images;
    }
    return undefined;
  }

  // ###Method - deferAllToLoadEvent() method to load all images with .lazy class after DOM Load event fires
  //
  // > parameters
  // >
  // + *event* - a native dom event
  // + *function* - optional callback( images ) invoked on image load completion with loaded images array
  //
  // > returns
  // >
  // + Nothing
  function deferAllToLoadEvent(event, cb) {
    const onResourcesLoaded = function () {
      _loadImages(document, cb);
      window.removeEventListener('load', onResourcesLoaded);
    };
    window.addEventListener('load', onResourcesLoaded);
  }

  module.exports = {
    loadContained,
    loadAll,
    deferAllToLoadEvent,
  };
}
