// @component/common/src/util/tags/AddToWishList.js
module.exports = function (response) {
  const tagManager = require('../TagManagerUtil');
  let price;
  let quantity;
  let utagData;
  let subtotal;
  if (response) {
    price = response.retailPriceWhenAdded ? response.retailPriceWhenAdded : '';
    quantity = response.selectedQty ? response.selectedQty : '';
    utagData = {
      event_name: 'wishlist add',
      product_id: [response.productId ? response.productId.toString() : ''],
      product_name: [response.productName],
      product_upc: [response.upcNumber],
      product_color: [response.color],
      product_size: [response.size],
      product_original_price: response.originalPrice ? [response.originalPrice.replace(/,/g, '')] : [],
    };

    if (price !== '' && quantity !== '') {
      price = parseFloat(price);
      quantity = parseInt(quantity, 10);
      if (!Number.isNaN(Number(price)) && !Number.isNaN(Number(quantity))) {
        subtotal = price * quantity;
        utagData.product_price = [price.toFixed(2)];
        utagData.product_sub_total = [subtotal.toFixed(2)];
        utagData.product_quantity = [quantity.toString()];
      }
    }

    tagManager.fireTag('link', utagData);
  }
};
