// @component/common/src/util/tags/CategoryPageEvents.js
module.exports = function (response) {
  const tagManager = require('../TagManagerUtil');
  if (response) {
    const data = (response.analytics || {}).data || {};
    const facetItems = [];
    const facetName = data.facetNames;
    const utagData = {
      event_name: 'product list updated',
      product_id: data.productID || [],
      facet_values: data.facetValues || [],
      product_rating: data.productRating || [],
      product_reviews: data.productReviews || [],
      results_current_page: data.resultsCurrentPage || '',
      results_per_page: data.resultsPerPage || '',
      sort_type: data.sortType || '',
      total_results: data.totalResults || '',
      product_placement_reason: data.productPlacementReason || '',
    };
    if (facetName && facetName.length) {
      let item;
      for (let i = 0; i < facetName.length; i++) {
        item = facetName[i].search('SIZE') > 0 ? `SIZE_TYPE:${facetName[i]}` : facetName[i];
        facetItems.push(item);
      }
    }
    utagData.facet_names = facetItems.length ? facetItems : [];
    tagManager.fireTag('view', utagData);
  }
};
