// @component/common/src/util/tags/PolarisAddToWishList.js
module.exports = function (response) {
  if (response) {
    const TagManagerUtil = require('../TagManagerUtil');
    const ObjectUtil = require('../ObjectUtil');
    const product = response.product || {};
    const upcs = product.relationships ? product.relationships.upcs : [];
    const respUpcId = response.upcId ? response.upcId.toString() : '';
    const selectedUpc = upcs.filter(upc => upc.identifier && upc.identifier.upcId === respUpcId);
    const selectedUpcObj = selectedUpc.length ? selectedUpc[0] : {};
    const upcNumber = selectedUpcObj.id ? selectedUpcObj.id : '';
    const tieredPrices = ObjectUtil.hasChildProperty(selectedUpcObj, 'pricing.price.tieredPrice') ? selectedUpcObj.pricing.price.tieredPrice : [];
    const getPrice = function (priceObj) {
      const values = priceObj.values;
      if (values && values.length && values[0].value) {
        return values[0].value.toString();
      }
      return '';
    };
    const orgPrice = getPrice(tieredPrices.length ? tieredPrices[0] : {});
    const price = tieredPrices.length > 1 ? getPrice(tieredPrices[tieredPrices.length - 1]) : orgPrice;
    const utagData = {
      event_name: 'wishlist add',
      product_id: [response.id ? response.id.toString() : ''],
      product_name: [product.detail ? product.detail.name : ''],
      product_price: [price],
      product_color: [response.currentColor || ''],
      product_size: [response.currentSize || ''],
      product_upc: [upcNumber.toString()],
      product_original_price: [orgPrice],
    };

    TagManagerUtil.fireTag('link', utagData);
  }
};
