// @component/common/src/util/tags/QuickView.js
module.exports = function (response, IShip, prosDetails) {
  const tagManager = require('../TagManagerUtil');
  if (response) {
    let productType;
    if (response.masterProduct) {
      productType = ['MASTER'];
    } else {
      productType = response.masterProductId === '' ? ['SINGLE ITEM'] : ['MEMBER'];
    }
    const colorwayPrice = response.colorwayPrice ? response.colorwayPrice : {};
    const utagData = {
      order_by_phone: response.phoneOnly ? 'true' : 'false',
      page_type: 'product quickview',
      event_name: 'product quickview',
      product_category_id: [response.categoryId],
      product_category_name: [response.typeName],
      product_id: [response.productID],
      product_name: [response.productDescription],
      product_type: productType,
      product_brand: [response.brand],
    };
    let origPrice = '';
    let salePrice = '';
    const iShip = IShip || {
      convertPrice(price) {
        return price;
      },
    };
    const cwOrigPrice = colorwayPrice.originalPrice ? iShip.convertPrice(colorwayPrice.originalPrice.toString()) : '';
    const cwOrigPriceHigh = colorwayPrice.originalPriceHigh ? ` - ${iShip.convertPrice(colorwayPrice.originalPriceHigh.toString())}` : '';
    const cwRetailPrice = colorwayPrice.retailPrice ? iShip.convertPrice(colorwayPrice.retailPrice.toString()) : '';
    const cwRetailPriceHigh = colorwayPrice.retailPriceHigh ? ` - ${iShip.convertPrice(colorwayPrice.retailPriceHigh.toString())}` : '';

    if (response.masterProduct) {
      origPrice = [cwOrigPrice + cwOrigPriceHigh];
      salePrice = [cwRetailPrice + cwRetailPriceHigh];
    } else {
      origPrice = [response.price ? iShip.convertPrice(response.price.toString()) : ''];
      salePrice = response.onSale ? [response.salePrice ? iShip.convertPrice(response.salePrice.toString()) : ''] : origPrice;
    }

    utagData.product_price = salePrice;
    utagData.product_original_price = origPrice;

    if (prosDetails) {
      utagData.event_name = prosDetails.eventName || 'recommendation link click';

      if (prosDetails.model) {
        utagData.recommendation_model = prosDetails.model;
      }

      if (prosDetails.position) {
        utagData.recommendation_position = prosDetails.position;
      }

      if (prosDetails.title) {
        utagData.recommendation_title = prosDetails.title;
      }

      if (prosDetails.zone) {
        utagData.recommendation_zone = prosDetails.zone;
      }
    }

    utagData.product_rating = [response.custRatings || '0.0'];

    utagData.product_reviews = [(response.numberOfReviews || 0).toString()];
    tagManager.fireTag('view', utagData);
  }
};
